### VARIANT-SPECIFIC DP AND GQ PLOTS  ###
### Argument 1: a file containing a tabular form of the variants characteristics (space-separate). This is normally generated by bcftool query.
###             These characteristics are needed: GQ_MEAN, DP
### Argument 2: Directory location where the plots should be saved 

library(ggplot2)
library(data.table)
options(warn=-1)

multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  library(grid)

  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)

  numPlots = length(plots)

  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                    ncol = cols, nrow = ceiling(numPlots/cols))
  }

 if (numPlots==1) {
    print(plots[[1]])

  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))

    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))

      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}


# Read arguments #
args <- commandArgs(trailingOnly = TRUE)
inD = as.character(args[1])
outD = as.character(args[2])
NIND = as.numeric(args[3])

#inD="/humgen/atgu1/fs03/wip/aganna/fin_seq/processed/seq/temp/G77318RH_table"
#outD="/humgen/atgu1/fs03/wip/aganna/fin_seq/processed/seq/plots/"
#NIND=597

# Read data
dnames <- read.table(paste0(inD,"_SNP_EX.txt"), header=F, comment.char = "&",nrows=1, stringsAsFactor=F)
a <- strsplit(as.character(dnames[1,]),"]")

print("Reading data")
d1 <- fread(paste0(inD,"_SNP_EX.txt"), header=F, stringsAsFactor=F)
d2 <- fread(paste0(inD,"_SNP_NEX.txt"), header=F, stringsAsFactor=F)
d3 <- fread(paste0(inD,"_INDEL_EX.txt"), header=F, stringsAsFactor=F)
d4 <- fread(paste0(inD,"_INDEL_NEX.txt"), header=F, stringsAsFactor=F)

colnames(d1) <- sapply(a[2:length(a)],"[[",2)
colnames(d2) <- sapply(a[2:length(a)],"[[",2)
colnames(d3) <- sapply(a[2:length(a)],"[[",2)
colnames(d4) <- sapply(a[2:length(a)],"[[",2)


print("Structuring data")

newdata1 <- data.frame(GQ=as.numeric(d1$GQ_MEAN), DP=as.numeric(d1$DP)/NIND)
newdata2 <- data.frame(GQ=as.numeric(d2$GQ_MEAN), DP=as.numeric(d2$DP)/NIND)
newdata3 <- data.frame(GQ=as.numeric(d3$GQ_MEAN), DP=as.numeric(d3$DP)/NIND)
newdata4 <- data.frame(GQ=as.numeric(d4$GQ_MEAN), DP=as.numeric(d4$DP)/NIND)



print("Now plotting")

png(paste0(outD,"DP_by_variant.png"), width=1200, height=800, type="cairo")

p1 <- ggplot(aes(x = DP), data=newdata1) + geom_histogram() +ylab("Density") + xlab("Average Sequencing Depth (truncated to 100)") + ggtitle("Average Sequencing Depth per variant - SNPs, EXOMES") + xlim(c(0,100)) + geom_vline(xintercept=mean(newdata1$DP),colour="yellow", linetype=2) + geom_vline(xintercept=median(newdata1$DP),colour="yellow", linetype=1)

p2 <- ggplot(aes(x = DP), data=newdata2) + geom_histogram() +ylab("Density") + xlab("Average Sequencing Depth (truncated to 100)") + ggtitle("Average Sequencing Depth per variant  - SNPs, NON EXOMES") + xlim(c(0,100)) + geom_vline(xintercept=mean(newdata2$DP),colour="yellow", linetype=2) + geom_vline(xintercept=median(newdata2$DP),colour="yellow", linetype=1)

p3 <- ggplot(aes(x = DP), data=newdata3) + geom_histogram() +ylab("Density") + xlab("Average Sequencing Depth (truncated to 100)") + ggtitle("Average Sequencing Depth per variant  - INDELs, EXOMES") + xlim(c(0,100)) + geom_vline(xintercept=mean(newdata3$DP),colour="yellow", linetype=2) + geom_vline(xintercept=median(newdata3$DP),colour="yellow", linetype=1)

p4 <- ggplot(aes(x = DP), data=newdata4) + geom_histogram() +ylab("Density") + xlab("Average Sequencing Depth (truncated to 100)") + ggtitle("Average Sequencing Depth per variant  - INDELs, NON EXOMES") + xlim(c(0,100)) + geom_vline(xintercept=mean(newdata4$DP),colour="yellow", linetype=2) + geom_vline(xintercept=median(newdata4$DP),colour="yellow", linetype=1)

multiplot(p1, p2, p3, p4, cols = 2)

dev.off()


png(paste0(outD,"GQ_MEAN_by_variant.png"), width=1200, height=800, type="cairo")

p1 <- ggplot(aes(x=GQ), data=newdata1) + geom_histogram() +ylab("Density") + xlab("Mean GQ (truncated to 100)") + ggtitle("Mean GQ per variant - SNPs, EXOMES") + xlim(c(0,100)) + geom_vline(xintercept=mean(newdata1$GQ),colour="yellow", linetype=2) + geom_vline(xintercept=median(newdata1$GQ),colour="yellow", linetype=1)

p2 <- ggplot(aes(x = GQ), data=newdata2) + geom_histogram() +ylab("Density") + xlab("Mean GQ (truncated to 100)") + ggtitle("Mean GQ per variant - SNPs, NON EXOMES") + xlim(c(0,100)) + geom_vline(xintercept=mean(newdata2$GQ),colour="yellow", linetype=2) + geom_vline(xintercept=median(newdata2$GQ),colour="yellow", linetype=1)

p3 <- ggplot(aes(x = GQ), data=newdata3) + geom_histogram() +ylab("Density") + xlab("Mean GQ (truncated to 100)") + ggtitle("Mean GQ per variant - INDELs, EXOMES") + xlim(c(0,100)) + geom_vline(xintercept=mean(newdata3$GQ),colour="yellow", linetype=2) + geom_vline(xintercept=median(newdata3$GQ),colour="yellow", linetype=1)

p4 <- ggplot(aes(x = GQ), data=newdata4) + geom_histogram() +ylab("Density") + xlab("Mean GQ (truncated to 100)") + ggtitle("Mean GQ per variant - INDELs, NON EXOMES") + xlim(c(0,100)) + geom_vline(xintercept=mean(newdata4$GQ),colour="yellow", linetype=2) + geom_vline(xintercept=median(newdata4$GQ),colour="yellow", linetype=1)


multiplot(p1, p2, p3, p4, cols = 2)

dev.off()


png(paste0(outD,"DP_by_GQ_by_variant.png"), width=1200, height=800, type="cairo")

p1 <- ggplot(data=newdata1,aes(x=GQ,y=DP)) + stat_binhex(colour="white",na.rm=TRUE) + xlab("GQ - limited to 500") + ylab("DP - limited to 200") + ylim(c(0,200)) + scale_fill_gradientn(colours=c("green1","red"),name = "Frequency",na.value=NA, trans = "log",guide=FALSE)+ theme_bw() + ggtitle("DP by GQ - SNPs, EXOMES") + stat_smooth(se = FALSE,lwd=2,span = 1) + xlim(c(0,500))

p2 <- ggplot(data=newdata2,aes(x=GQ,y=DP)) + stat_binhex(colour="white",na.rm=TRUE) + xlab("GQ - limited to 500") + ylab("DP - limited to 200") + ylim(c(0,200)) + scale_fill_gradientn(colours=c("green1","red"),name = "Frequency",na.value=NA, trans = "log",guide=FALSE)+ theme_bw() + ggtitle("DP by GQ - SNPs, NON EXOMES") + stat_smooth(se = FALSE,lwd=2,span = 1) + xlim(c(0,500))

p3 <- ggplot(data=newdata3,aes(x=GQ,y=DP)) + stat_binhex(colour="white",na.rm=TRUE) + xlab("GQ - limited to 500") + ylab("DP - limited to 200") + ylim(c(0,200)) + scale_fill_gradientn(colours=c("green1","red"),name = "Frequency",na.value=NA, trans = "log",guide=FALSE)+ theme_bw() + ggtitle("DP by GQ - INDELs, EXOMES") + stat_smooth(se = FALSE,lwd=2,span = 1) + xlim(c(0,500))

p4 <- ggplot(data=newdata4,aes(x=GQ,y=DP)) + stat_binhex(colour="white",na.rm=TRUE) + xlab("GQ - limited to 500") + ylab("DP - limited to 200") + ylim(c(0,200)) + scale_fill_gradientn(colours=c("green1","red"),name = "Frequency",na.value=NA, trans = "log",guide=FALSE)+ theme_bw() + ggtitle("DP by GQ - INDELs, NON EXOMES") + stat_smooth(se = FALSE,lwd=2,span = 1) + xlim(c(0,500))


multiplot(p1, p2, p3, p4, cols = 2)

dev.off()





